File: zfs_get_quota.rb
#!/usr/bin/env ruby

require 'rubygems'
require 'etc'
require 'csv'
require 'json'
require 'yaml'
require 'optparse'
require 'logger'
require 'terminal-table'
require 'filesize'
require 'pp'
require 'thor'

class ZFSQuotaReporter
  def initialize(quotatype, inputfile)
    @quotatype = quotatype
    @inputfile = inputfile
    
    case quotatype
    when 'user'
      @table_headers = ['User','UID','Space','mtime']
    when 'group'
      @table_headers = ['Group','GID','Space','mtime']
    end
  end

  def execute!
    entries = []
    rows = []
    total = 0

    if valid_yaml_file?(@inputfile)
      entries = parse_yaml(@inputfile)
    else
      puts "Must provide valid YAML file"
      exit 1
    end

    entries.sort_by { |h| h[:space] }.each do |e|
      total = total + e[:space]
      a = [e[:name],e[:id],Filesize.from("#{e[:space]} B").pretty,e[:mtime]]
      rows << a
    end

    table = Terminal::Table.new(:headings => @table_headers, :rows => rows)

    table.add_separator
    table.add_row(['Total', '---', Filesize.from("#{total} B").pretty, '---'])

    puts table
  end

  private

  def valid_yaml_file?(file)
    begin
      contents = YAML.load(File.open(file))
      return false unless contents.is_a?(Array)
    rescue Exception => e
      STDERR.puts e.message
      return false
    else
      return true
    end
  end

  def parse_yaml(file)
    contents = YAML.load(File.open(file))
    contents
  end
end

class ZFSQuotaParser
  def initialize(quotatype, inputfile, output, format)
    @quotatype = quotatype
    @inputfile = inputfile
    @output = output
    @format = format
    
    case quotatype
    when 'user'
      @get_id_method = 'getpwuid'
    when 'group'
      @get_id_method = 'getgrgid'
    end
  end

  def execute!
    entries = process_inputfile
    sorted_entries = entries.sort_by { |e| e[:id] }.flatten
    save_output(sorted_entries)
  end

  private

  def process_inputfile
    entries = []

    # Get mtime of input file and convert to epoch
    mtime = File.mtime(@inputfile).strftime("%s").to_i
    f = File.open(@inputfile, "r")
    lines = f.readlines

    lines.each do |line|
      # Match each line that contains UID BYTES
      if line =~ /^([0-9]+)\s+([0-9]+)$/
        id = $1.to_i
        space = $2.to_i

        # Find existing entries else create new entry
        if e = entries.find { |e| e[:id] == id }
          # For existing entries, add the current space used to new value
          # This allows sum of space across multiple input files
          cur = e[:space]
          e[:space] = cur.to_i + space.to_i
        else
          begin
            name = Etc.send(@get_id_method, id).name
          rescue ArgumentError
            next
          end
          e = {}
          e[:name] = name
          e[:id] = id
          e[:space] = space
          e[:mtime] = mtime
          entries << e
        end
      end
    end

    entries
  end

  def save_output(entries)
    case @format
    when 'csv'
      CSV.open(@output, "w") do |csv|
        entries.each do |e|
          csv << [e[:id],e[:name],e[:space],e[:mtime]]
        end
      end
    when 'yaml'
      File.open(@output, "w") { |f| f.write entries.to_yaml }
    end
  end
end

class ZFSGetQuota < Thor
  method_option :quotatype, :default => 'user', :aliases => "-t", :desc => "Type of quota output to parse. Valid options are 'user' and 'group'."
  method_option :output, :default => '/tmp/zfs_userspace.yaml', :aliases => "-o", :desc => "Where to output quota report"
  method_option :inputfile, :required => true, :aliases => "-i", :desc => "File from running 'zfs userspace'"
  method_option :format, :default => 'yaml', :aliases => "-f", :desc => "Output format"
  desc 'parse', "Parse raw quota input files"
  def parse
    validate_quotatype(options.quotatype, 'parse')
    @parser = ZFSQuotaParser.new(options.quotatype, options.inputfile, options.output, options.format)
    @parser.execute!
  end

  method_option :quotatype, :default => 'user', :aliases => "-t", :desc => "Type of quota to report.  Valid options are 'user' and 'group'."
  method_option :inputfile, :required => true, :aliases => "-i", :desc => "File from running parser"
  desc 'report', "Print quota reports"
  def report
    validate_quotatype(options.quotatype, 'report')
    @reporter = ZFSQuotaReporter.new(options.quotatype, options.inputfile)
    @reporter.execute!
  end

  private

  def validate_quotatype(quotatype, command_name)
    if quotatype !~ /^(user|group)$/
      puts "Invalid quotatype: #{quotatype}"
      help(command_name)
      exit 1
    end
  end
end

ZFSGetQuota.start

File: bin/zfs_get_quota.py
#!/usr/bin/env python

import argparse
import csv
import grp
import json
import os
import prettytable
import pwd
import re
import sys

YAML_SUPPORT = True
try:
    import yaml
except ImportError, e:
    if str(e) != "No module named yaml":
        raise
    else:
        YAML_SUPPORT = False

if YAML_SUPPORT:
    SUPPORTED_FORMATS = ["csv", "json", "yaml"]
else:
    SUPPORTED_FORMATS = ["csv", "json"]

def parse(args, parser):
    if args.output == None:
        args.output = "/tmp/fhgfs_%sspace.%s" % (args.quotatype, args.format)
    print args

    entries = []
    mtime = int(os.stat(args.input).st_mtime)
    lines = []
    with open(args.input) as f:
        lines = f.readlines()
    for line in lines:
        m = re.search('^([0-9]+)\s+([0-9]+)$', line)
        if not m:
            continue
        _id = int(m.group(1))
        _space = int(m.group(2))
        e = None
        for entry in entries:
            if entry["id"] == _id:
                e = entry
        if e:
            cur = e["space"]
            e["space"] = int(cur) + _space
        else:
            try:
                if args.quotatype == "group":
                    group = grp.getgrgid(_id)
                    name = group.gr_name
                elif args.quotatype == "user":
                    user = pwd.getpwuid(_id)
                    name = user.pw_name
                else:
                    name = None
            except KeyError:
                continue

            e = {}
            e["name"] = name
            e["id"] = _id
            e["space"] = _space
            e["mtime"] = mtime
            entries.append(e)
    sorted_entries = sorted(entries, key=lambda k: k["id"])

    if args.format == "csv":
        with open(args.output, "w") as csvfile:
            csvwriter = csv.writer(csvfile, delimiter=",")
            for entry in sorted_entries:
                csvwriter.writerow([entry["id"], entry["name"], entry["space"], entry["mtime"]])
    elif args.format == "json":
        with open(args.output, "w") as jsonfile:
            json.dump(sorted_entries, jsonfile, sort_keys=True, indent=4)
    elif args.format == "yaml":
        with open(args.output, "w") as yamlfile:
            yaml.dump(sorted_entries, yamlfile, default_flow_style=False)


def sizeof_fmt(num, suffix="B"):
    for unit in ["", "K", "M", "G", "T", "P"]:
        if abs(num) < 1024.0:
            return "%3.2f%s%s" % (num, unit, suffix)
        num /= 1024.0
    return "%.2f%s%s" % (num, "Y", suffix)


def report(args, parser):
    table_headers = ["Name", "ID", "Space", "mtime"]

    filename, ext = os.path.splitext(os.path.basename(args.input))
    entries = None
    if ext == ".csv":
        entries = []
        with open(args.input) as csvfile:
            csvreader = csv.reader(csvfile, delimiter=",")
            for row in csvreader:
                entry = {}
                entry["id"] = row[0]
                entry["name"] = row[1]
                entry["space"] = int(row[2])
                entry["mtime"] = row[3]
                entries.append(entry)
    elif ext == ".json":
        with open(args.input) as jsonfile:
            entries = json.load(jsonfile)
    elif ext == ".yaml":
        with open(args.input) as yamlfile:
            entries = yaml.load(yamlfile)
    else:
        print "Unknown file format: %s" % ext
        sys.exit(1)

    table = prettytable.PrettyTable(table_headers)
    table.hrules = prettytable.FRAME
    sorted_entries = sorted(entries, key=lambda k: k["space"])
    total_space = 0
    for entry in sorted_entries:
        total_space = total_space + entry["space"]
        space = sizeof_fmt(entry["space"])
        table.add_row([entry["name"], entry["id"], space, entry["mtime"]])
    table.add_row(["","","",""])
    table.add_row(["Total", "---", sizeof_fmt(total_space), "---"])
    print table

parser = argparse.ArgumentParser()
subparsers = parser.add_subparsers(dest='mode')
parser_parse = subparsers.add_parser('parse')
parser_report = subparsers.add_parser('report')

parser_parse.add_argument("-i", "--inputfile", help="Input file", dest="input", required=True)
parser_parse.add_argument("-t", "--quotatype", help="Type of quota to parse", choices=["user", "group"], dest="quotatype", default="user")
parser_parse.add_argument("-o", "--output", help="Where to output quota report", dest="output")
parser_parse.add_argument("-f", "--format", help="Output format", choices=SUPPORTED_FORMATS, dest="format", default="json")
parser_parse.set_defaults(func=parse)

parser_report.add_argument("-i", "--inputfile", help="Input file", dest="input", required=True)
parser_report.set_defaults(func=report)

args = parser.parse_args()
args.func(args, parser)

